Step3 : 리뷰 요청드립니다.

Mo-bile opened the pull request:
안녕하세요?
리뷰 요청드립니다 👍 

금번 보너스 번호 기능을 추가하면서 어디 object에 2,3등 간에 등수를 결정시킬지 고민을 오래한것 같습니다.

그 결과
1. 로또 번호에 대한 추가 검증이므로 'rankDecideByBonusNumber 메서드를 추가
2. `Bonus` 클래스를 새로 만들어 보너스 번호와 로또 번호의 한번 더 검증
3. `Rank` 클래스에 2 혹은 3등인지 구분하는 메서드

이렇게 크게 구분하여 이번 미션을 진행한 것 같습니다.
단순하게 구현하면 indent 가 2이상 늘어나버리고 복잡한 형태가 만들어지고, 보기싫어지고 읽기 싫어지더라고요

처음에 요구사항을 보면 쉽다고 생각했는데 막상 해보니 녹록치 않은 과제였던 것 같습니다

좋은 과제 감사드립니다 💯

Commit:
refactor : findMatchCount 의 인자 타입을 변경하여 테스트 쉽게 변경

Commit:
refactor
1. 가변인자 변경
2. 일부 메서드 명 변경

Commit:
refactor : 가변인자 개선

Commit:
feat : Match 상수를 추가하여 추가된 인자 처리

Commit:
fix : 출력문 수정

Commit:
refactor
1. 메서드 호출을 명확히 하기위해 호출 변경
2. Lotto 호출 생성자 방식에 맞춰 테스트 코드 픽스쳐 변경

Commit:
refactor : Match enum을 Rank 로 변경하여 객체의 존재(what he is) 명확화

Commit:
feat : 보너스 번호 관리를 위한 Bonus 객체 추가

Commit:
refactor : Rank 내 멤버와 메서드명 등을 명확히 변경하여 그 행동을 드러냄

Commit:
feat : 로또 5개일치시 2, 3등 보너스 번호로 구분하여 등수 선택 기능 구현 완료

Commit:
doc : 체크리스트 완료

Commit:
refactor : 최종 2,3 등 등수결정에 관한 책임을 `Lotto` 객체로 이관하여 Lotto 번호에 대한 관리 책임 강화

Commit:
fix : 2개 이하로 일치할 경우 match되는게 없어서 발생하는 NPE 방지 에러 해결

Commit:
refactor : 불필요한 increaseCount 메서드 제거

javajigi commented on src/main/java/lotto/controller/LottoApplication.java:
```diff
@@ -14,7 +14,8 @@ public static void main(String[] args) {
         String winningNumbers = InputView.inputWinningNumbers();
-        WinningResult winningResult = lottoGame.showWinners(winningNumbers);
+        int bonusNumbers = InputView.inputBonusNumbers();
+        WinningResult winningResult = lottoGame.displayWinners(winningNumbers, bonusNumbers);
```
AI 피드백을 받아 더 적합한 메서드 이름으로 리팩터링 해보면 어떨까?
메서드가 하는 일과 displayWinners는 적합해 보이지 않음.

javajigi commented on src/main/java/lotto/domain/model/Bonus.java:
```diff
@@ -0,0 +1,34 @@
+
+public class Bonus {
+    
+    int bonusNumber;
```
```suggestion
    private final int bonusNumber;
```

javajigi commented on src/main/java/lotto/domain/model/Bonus.java:
```diff
@@ -0,0 +1,34 @@
+    
+    public Bonus() {
+        this(0);
+    }
```
불필요한 생성자이지 않을까?

javajigi commented on src/main/java/lotto/domain/model/Bonus.java:
```diff
@@ -0,0 +1,34 @@
+            }
+        }
+        return false;
+    }
```
lottoNumbers.cotains(bonus);

오히려 거꾸로 메시지를 보내는 것은 어떨까?

javajigi commented on src/main/java/lotto/domain/model/LottoNumber.java:
```diff
@@ -0,0 +1,34 @@
+    }
+    
+    private void validate(int bonusNumber) {
+        if(bonusNumber < 1 || bonusNumber > 45) {
```
Lotto의 int 값도 모두 같은 속성을 가지고 있다.
Lotto의 int 값을 포장하는 객체와 Bonus를 합쳐 함께 사용할 수 있는 객체로 구현해 보면 어떨까?

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -58,17 +65,27 @@ private static List<Integer> generateNumberList() {
     }
     
-    public int findMatchCount(List<Integer> winNumbers) {
+    public int findMatchCount(Lotto winNumbers) {
```
```suggestion
    private int findMatchCount(Lotto winNumbers) {
```
private으로 구현해도 되지 않을까?

Mo-bile replied:
@javajigi 
처음에 이 구현할 때는 메서드의 접근제어자를 public으로 이용하면서 테스트 코드를 사용했습니다
그러다가 같은 객체 내 다른 메서드 안에서 사용하게 되어 private로 바꿔야 하게 되었습니다
그러나 private 로 하면 테스트코드를 쓸수가 없게 되어서 public으로 두었습니다.
일단 private 로 바꾸도록 할게요 👍

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -58,17 +65,27 @@ private static List<Integer> generateNumberList() {
         return matchCount;
     }
+    
+    public Rank rankDecideByBonusNumber(Bonus bonus, Rank rank) {
```
Rank를 수동적인 객체로 바라보는 느낌
Rank가 등수를 결정하도록 구현해 보면 어떨까?

javajigi commented on src/main/java/lotto/domain/model/WinningResult.java:
```diff
@@ -2,34 +2,47 @@
+        return map;
+    }
+    
+    private static List<Integer> winningNumberArgsBuilder(int[] ints) {
```
AI에서 array를 List로 변환으로 검색해 볼 것을 추천

javajigi commented on src/main/java/lotto/domain/model/Lottos.java:
```diff
@@ -17,10 +17,10 @@ private static List<Lotto> generateLottos(int num) {
     }
     
-    public WinningResult findWinners(Lotto winnerLotto) {
+    public WinningResult identifyWinners(Lotto winnerLotto, Bonus bonusNumber) {
```
Lotto와 Bonus를 가지는 당첨 번호라는 객체를 분리해 보면 어떨까?

javajigi commented on src/test/java/lotto/domain/model/LottosTest.java:
```diff
@@ -16,10 +16,20 @@ class LottosTest {
-        WinningResult exptedwinner = new WinningResult(1, 0, 0, 0);
-        assertThat(winningResult).isEqualTo(exptedwinner);
+        WinningResult winningResult = lottos.identifyWinners(new Lotto(35, 36, 37, 43, 44, 45), new Bonus(1));
+        WinningResult expectedWinner = new WinningResult(1, 1, 0, 0, 0, 0);
```
```suggestion
        WinningResult expectedWinner = new WinningResult();
        expectedWinner.increaseRankCount(Rank.FIRST);
        expectedWinner.increaseRankCount(Rank.SECOND);
```

생성자를 통해 등수를 전달하니 몇 등인지를 파악하는데 어려움이 있을 것 같다.
위와 같이 구현하는 것은 어떨까?

javajigi commented on src/test/java/lotto/domain/model/LottoTicketsTest.java:
```diff
@@ -16,10 +16,20 @@ class LottosTest {
-        assertThat(winningResult).isEqualTo(exptedwinner);
+        WinningResult winningResult = lottos.identifyWinners(new Lotto(35, 36, 37, 43, 44, 45), new Bonus(1));
+        WinningResult expectedWinner = new WinningResult(1, 1, 0, 0, 0, 0);
+        assertThat(winningResult).isEqualTo(expectedWinner);
```
피드백 남기다보니 increaseRankCount 이름도 뭔가 의도를 파악하기 힘듦
AI를 통해 피드백을 받아보고 개선해 봤으면 좋겠음

javajigi commented on src/test/java/lotto/domain/model/LottoTest.java:
```diff
@@ -22,43 +21,54 @@ class LottoTest {
         Lotto lotto = new Lotto(40, 41, 42, 43, 44, 45);
-        assertThat(lotto.findMatchCount(List.of(35, 36, 37, 43, 44, 45)))
+        assertThat(lotto.findMatchCount(new Lotto(35, 36, 37, 43, 44, 45)))
             .isEqualTo(3);
```
```suggestion
        assertThat(lotto.findMatchCount(new Lotto(35, 36, 37, 43, 44, 45)))
.isEqualTo(Rank.FIFTH);
```
숫자가 아니라 Rank로 비교하면 어떨까?

javajigi reviewed (changes requested):
3단계 미션 진행하느라 수고했어요. 👍 
적은 수의 객체 분리로 로또 미션 진행했는데요.

- 규칙 3: 모든 원시값과 문자열을 포장한다.
- 규칙 8: 일급 콜렉션을 쓴다.

위 두 개 원칙을 지키면서 객체 분리에 대한 리팩터링 도전해 보세요.
추가로 메서드 이름이 의도를 드러내지 못하는 경우가 많은데요.
메서드 이름은 가능하면 AI에게 피드백을 받아 개선해 보는 습관을 들여보면 좋겠습니다.

Commit:
refactor : Bonus 에 필드 불변으로 변경

Commit:
feat : match Count 로 테스트 코드 변경

Commit:
refactor
1. Lotto 와 Bonus 간에 메시지를 주는 객체간 관계 재조정 및 테스트 제거
2. findMatchCount 접근제어자 private 로 변경하면서 테스트 제거

Commit:
refactor : rankDecideByBonusNumber 메서드 Lotto 에서 Rank 객체로 이관하여 Rank의 능동성을 좀 더 드러나게 변경

Commit:
refacotr : 가변배열 생성자 List로 전환방식 반복문에서 스트림 방식으로 방식하여 리턴하는 방식으로 변경

Commit:
refactor : 주요 필드명, 인자명, 메서드명변경하여 의도 명확히 드러냄

Commit:
refactor : WinningResult 의 가변생성자 방식 불필요로 제거하고 후 recordRank 이용하여 테스트 픽스쳐 가독성향상

Commit:
refactor : 기존 Bonus 객체를 LottoNumber 로 객체 명과 책임을 변경한다
그 이유는 Bonus 의 역할과 지정된 행동이 일반적인 로또 번호와 차이가 없음

Commit:
refactor : 로또 관련 number, lotto, lottos 3가지의 객체명과 필드명 재정립하여 도메인의 역할 드러나게 변경

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -23,14 +23,18 @@ public Lotto(String lottoNumbers) {
+        List<LottoNumber> numbers = new ArrayList<>();
         for(String number: extractWinnerLottoNumber(lottoNumbers)) {
-            numbers.add(Integer.parseInt(number));
+            numbers.add(new LottoNumber(Integer.parseInt(number)));
```
LottoNumber에 String 타입을 가지는 생성자가 있다면 LottoNumber를 생성하는 개발자가 타입 변환을 하지 않고 편리하게 사용할 수 있지 않을까?

javajigi commented on src/main/java/lotto/domain/model/Lotto.java:
```diff
@@ -39,36 +43,51 @@ private static String[] extractWinnerLottoNumber(String winnerLottoNumber) {
+            if(lottoNumber.isSameNumber(number.value())) {
+                return true;
+            }
+        }
+        return false;
+    }
```
```suggestion
    public boolean isContain(LottoNumber lottoNumber) {
        return this.numbers.contains(lottoNumber);
    }
```
콜렉션 api를 적극 활용해 본다.

javajigi commented on src/main/java/lotto/domain/model/LottoTickets.java:
```diff
@@ -17,10 +17,10 @@ private static List<Lotto> generateLottos(int num) {
     }
     
-    public WinningResult findWinners(Lotto winnerLotto) {
+    public WinningResult identifyWinners(Lotto winnerLotto, LottoNumber bonusNumber) {
```
클린 코드 책에 함수 인수와 관련해 다음과 같은 가이드가 있다.
인수의 수를 줄일 수 있는 방법을 찾아본다.

## 함수 인수
- 함수에서 이상적인 인수 개수는 0개(무항)이다. 다음은 1개이고, 다음은 2개이다.
- 3개는 가능한 피하는 편이 좋다.
- 4개 이상은 특별한 이유가 있어도 사용하면 안된다

javajigi commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,46 @@
+        this.winnerReturn = winnerReturn;
+    }
+    
+    public static Rank fromLottoNumber(int matchCount) {
```
```suggestion
    public static Rank fromLottoNumber(int matchCount, boolean bonusMatch) {
```
이렇게 값이 전달되야 2등을 구할 수 있지 않을까?

javajigi commented on src/test/java/lotto/domain/model/LottoTest.java:
```diff
@@ -13,52 +15,28 @@ class LottoTest {
-        assertThat(lotto.findMatchCount(List.of(42, 36, 37, 43, 44, 45)))
-            .isEqualTo(4);
+    @ParameterizedTest
+    @ValueSource(ints = {0,46})
```
```suggestion
    @ValueSource(ints = {0, 46})
```

javajigi commented on src/test/java/lotto/domain/model/WinningResultTest.java:
```diff
@@ -2,40 +2,49 @@
-        WinningResult winningResult = new WinningResult(0, 1, 0, 0);
+        WinningResult winningResult = new WinningResult();
+        winningResult.recordRank(Rank.FOURTH);
         assertThat(winningResult.calculateTotalReturn(14000)).isEqualTo("3.57");
```
돈을 나타내는 숫자의 경우 천 단위를 구분하기 위해 "1_000"와 같이 언더바를 통해 구분할 수 있다.

javajigi commented on src/test/java/lotto/domain/RankTest.java:
```diff
@@ -0,0 +1,31 @@
+    
+    @Test
+    public void 랭크3등인데_보너스_번호를_맞아서_2등을_결정한다() {
+        assertThat(Rank.THIRD.rankDecideByBonusNumber(new Lotto(40, 41, 42, 43, 44, 45), new LottoNumber(40))).isEqualTo(Rank.SECOND);
```
굳이 2등을 분리해서 구해야 하나?
하나의 메서드에서 2등, 3등을 함께 결정할 수 있어야 하지 않을까?

javajigi reviewed (changes requested):
피드백 반영하느라 수고 했어요. 👍 
2단계에서 3단계로 오면서 당첨 번호에 보너스 번호가 추가되었는데요.
당첨 번호를 담당하는 객체를 추가해 보면 어떨까요?
이렇게 당첨 번호를 담당하는 객체를 추가하면서 Lotto의 책임과 역할을 당첨 번호를 담당하는 객체로 이동해 볼 것을 추천합니다.

Commit:
refactor : refactor : 컨벤션 정렬 및 안쓰는 import 정리

Commit:
1. LottoNumber 의 생성자에 문자열 입력도 가능하게 변경
2. 포함여부에 대한 구현 대신 컬렉션 API 활용하여 간편하게 변경
3. LottoNumber 객체가 생겨서 출력이 상이해진 부분 수정
4. 테스트 코드 금액표시에 3자리수 마다 _ 추가하여 가독성 향상

Commit:
refactor
1. WinningLotto 신설해서 당첨값 관리 책임에 대한 객체 신설
2. Rank 와 Lotto 간에 순환 참조 개선 (순위 뽑은 뒤 다시 2,3등 뽑는구조 개선)

Commit:
refactor
1. 등수를 최종 지정하는 역할을 Rank 에서 WinningLotto 로 이관

Commit:
refactor : 랭킹 결정하는 메서드 `Rank` 로 이관하여 규칙을 아는 자가 할수있게 변경하고 `WinningLotto` 는 로또번호 lotto 객체와 Rank 사이에 조율할 수있게 변경

Commit:
refactor : 생성자 리팩터링

Commit:
refactor : 각종 컨벤션 준수하게 리팩터링 및 정리

Commit:
refactor : 반복문을 stream API 로 변경

Commit:
doc : 체크리스트 업데이트

Commit:
chore : 안쓰는 패키지 제거

Commit:
feat : 타당성 검증 코드 추가, 방어적 복사 추가

javajigi commented on src/main/java/lotto/domain/model/LottoNumber.java:
```diff
@@ -0,0 +1,18 @@
+package lotto.domain.model;
+
+public record LottoNumber(int value) {
```
## 상황
- 1만명의 사용자가 동시에 당첨 여부를 확인할 수 있어야 한다.
- 1명의 사용자는 평균 5장의 로또를 구매한 상태이다.

위 요구사항을 서버에서 생성되는 LottoNumber의 인스턴스의 갯수는?
1 * 5장 * 6개 숫자/1장 * 1만명 = 30만개이다.

동시에 생성되는 인스턴스 갯수가 너무 많다.
인스턴스 갯수를 줄일 수 있는 방법은?
로또 숫자 값을 LottoNumber 객체로 래핑하는 경우 매번 인스턴스가 생성되기 때문에 인스턴스의 갯수가 너무 많아져 성능이 떨어질 수 있다.
LottoNumber 인스턴스를 생성한 후 재사용할 수 있도록 구현한다.

힌트 : Map과 같은 곳에 인스턴스를 생성한 후 재사용하는 방법을 찾아본다.

javajigi commented on src/main/java/lotto/domain/model/WinningLotto.java:
```diff
@@ -0,0 +1,41 @@
+
+import lotto.domain.Rank;
+
+public record WinningLotto(LottoNumber bonusNumber, Lotto winNumbers) {
```
```suggestion
public record WinningLotto(Lotto winNumbers, LottoNumber bonusNumber) {
```
저 같으면 로또를 첫 번째 인자로 구현할 것 같음

javajigi commented on src/main/java/lotto/domain/model/WinningLotto.java:
```diff
@@ -0,0 +1,41 @@
+    
+    public WinningLotto {
+        valid(winNumbers, bonusNumber);
+    }
```
여러 생성자로 구현한 점 👍 
단, 당첨 번호의 경우 로또와 보너스 번호는 필수인 만큼 이 두 개의 생성자는 필요없지 않을까?

javajigi commented on src/main/java/lotto/domain/model/WinningLotto.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+    
+    public Rank rankDecide(Lotto lotto) {
```
WinningLotto로 이 메서드를 이동한 점 👍

javajigi commented on src/main/java/lotto/domain/model/WinningResult.java:
```diff
@@ -42,30 +30,23 @@ public String calculateTotalReturn(int pay) {
-    
-    private void increaseCount(Match match) {
-        this.matchMap.compute(match, (k, count) -> {
+    public void recordRank(Rank rank) {
```
👍

javajigi commented on src/main/java/lotto/domain/Rank.java:
```diff
@@ -0,0 +1,46 @@
+        this.winnerReturn = winnerReturn;
+    }
+    
+    public static Rank valueOf(int matchCount, boolean bonusMatch) {
```
👍

javajigi reviewed (changes requested):
이전 단계보다 확실히 코드가 많이 개선되었음을 느낄 수 있네요. 👍 
코드 컨벤션도 잘 지키고 있네요.
다음 단계 넘어가기 전에 꼭 연습해 봤으면 하는 부분이 있어 피드백 남겨요.
LottoNumber 재사용 부분만 반영하고 다음 단계로 가시죠.

Commit:
refacotor : 불필요한 생성자 제거

Commit:
feat : `lottoNumber`의 무한한 인스턴스 생성을 방지하기 위해 `lottoNumber` 의 캐싱하는 기능을 추가하여 인스턴스 생성을 방지하여 성능 개선

Commit:
docs : 리팩터링 목표 기록

Commit:
refactor : 상수값 추가

javajigi commented on src/main/java/lotto/domain/model/LottoNumber.java:
```diff
@@ -0,0 +1,18 @@
+
+public record LottoNumber(int value) {
+    
+    public LottoNumber(String value) {
```
캐시 적용 👍 
단 이와 같이 구현할 경우 캐시를 사용하지 않고 LottoNumber를 직접 생성 가능하다.
LottoNumber를 외부에서 직접 생성하지 못하도록 private으로 구현하고 캐시 효과를 내도록 리팩터링해본다.

힌트: 정적 팩터리 메서드

javajigi reviewed (approved):
캐시 적용 잘 했네요. 👍 
단, 캐시 적용에 아쉬운 부분이 있어 피드백했으니 4단계 진행할 때 함께 반영해 보세요.
4단계는 최대한 컴파일 에러를 내지 않으면서 안정적으로 기능 추가하는 연습해 볼 것을 추천해요.
